connectors/preflight_check.py

```diff
@@ -40,22 +41,98 @@ async def run(self):
         try:
             logger.info("Running preflight checks")
             self.running = True
-            if not (await self.es_management_client.wait()):
-                logger.critical(
-                    f"{self.elastic_config['host']} seems to be unreachable. Bye!"
-                )
-                return False
+
+            success, is_serverless = await self._check_es_server()
+            if not success:
+                return False, is_serverless
 
             await self._check_local_extraction_setup()
 
             valid_configuration = self._validate_configuration()
             available_system_indices = await self._check_system_indices_with_retries()
-            return valid_configuration and available_system_indices
+            return (valid_configuration and available_system_indices), is_serverless
         finally:
             self.stop()
             if self.es_management_client is not None:
                 await self.es_management_client.close()
 
+    async def _check_es_server(self):
+        """
+        Returns two values:
+
+        1. Whether or not the ES server is compatible with the running Connectors version
+        2. Whether or not the ES server is serverless
+        """
+
+        es_info = await self.es_management_client.wait()
+        if es_info is None:
+            logger.critical(
+                f"{self.elastic_config['host']} seems to be unreachable. Bye!"
+            )
+            return False, False
+
+        is_serverless = es_info["version"]["build_flavor"] == "serverless"
+        if is_serverless:
+            logger.info(
+                "Elasticsearch server is serverless, skipping version compatibility check."
+            )
+            return True, is_serverless
+
+        versions_compatible = await self._versions_compatible(
+            es_info["version"]["number"]
+        )
+        return versions_compatible, is_serverless
+
+    async def _versions_compatible(self, es_version):
+        """
+        Versions are incompatible in the following scenarios:
+
+        1. the major version is different
+            e.g. ES v7.13.0 vs Connectors v8.13.0.0
+        2. the ES minor is lower than Connectors
+            e.g. ES v8.12.0 vs Conectors v8.13.0.0
+
+        If the above is satisfied but the versions are not exactly the same we log a warning
+        """
+
+        # array legend: 0 - major, 1 - minor, 2 - patch
+        es_version_parts = list(
+            map(int, es_version.replace("-SNAPSHOT", "").split("."))
+        )
+        connector_version_parts = list(map(int, self.version.split(".")))
+
+        # major
+        if es_version_parts[0] != connector_version_parts[0]:
+            logger.critical(
+                f"Elasticsearch {es_version} and Connectors {self.version} are incompatible: major versions are different"
+            )
+            return False
+
+        # minor
+        if es_version_parts[1] > connector_version_parts[1]:
```

Comment:
Should this check for non-equality, rather than only if ES is greater?

Like if connectors is a higher version, and needs an ES Connector API that doesn't exist in the targeted ES deployment, we'll have issues.

Reply:
Actually, I see below that you check for `>` separately. I think we could just check both together, and error that they're non-equal. Could probably even do that with the major version check.

Also, the docstring says you only check that the ES minor isn't greater.

Reply:
> Like if connectors is a higher version, and needs an ES Connector API that doesn't exist in the targeted ES deployment, we'll have issues.

This is a good point. I think we could have it a bit different from the Kibana compatibility matrix as we have this dependency to consider.

Reply:
> Also, the docstring says you only check that the ES minor isn't greater.

I had a line at the end stating `If the above is satisfied but the versions are not exactly the same we log a warning` which I thought could be a catch-all for other inconsistencies, but I can see it just caused confusion.

Reply:
Actually I had implemented this wrong.

If the ES minor is greater than Connectors it should terminate. If the ES minor is lower than Connectors it should output a warning. I had these the wrong way around.

I'd prefer to keep them separate so we have the same compatibility as Kibana does with ES.

Reply:
👍 I suppose that makes sense. It assumes that Elasticsearch will be backwards-compatible on a minor upgrade, which is a reasonable assumption.

Reply:
wait, no, that's the opposite of what I said. 😕 
Where does Kibana's compatibility get defined? I'm surprised that an older kibana can't talk to a newer Elasticsearch.

Reply:
@jedrazb on the issue is a version compatibility matrix and reference to it being similar to Kibana but there's no linnk, do you remember where the information came from?

Reply:
@seanstory the version compatibility table is on Kibana's main README: https://github.com/elastic/kibana?tab=readme-ov-file#version-compatibility-with-elasticsearch

Reply:
> If the ES minor is greater than Connectors it should terminate. If the ES minor is lower than Connectors it should output a warning

This is the _opposite_ of what's in that table. 

<img width="792" alt="Screenshot 2024-06-26 at 1 14 31 PM" src="https://github.com/elastic/connectors/assets/5288246/a59ee733-966d-448c-b34d-50eeb9d1fc88">

Reply:
I think in my mind I was treating "newer" == "lower number" like its an age 🤦🏻

Reply:
I fixed it and changed the wording to use `older` and `newer` so it's easier to mentally parse compared to the table.